<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Библиотеки" data-toc-modified-id="Библиотеки-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Библиотеки</a></span></li><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Подготовка данных</a></span></li><li><span><a href="#Промежуточный-вывод" data-toc-modified-id="Промежуточный-вывод-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Промежуточный вывод</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span><ul class="toc-item"><li><span><a href="#Задача" data-toc-modified-id="Задача-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Задача</a></span></li><li><span><a href="#Решение" data-toc-modified-id="Решение-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Решение</a></span></li><li><span><a href="#Промежуточный-вывод" data-toc-modified-id="Промежуточный-вывод-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Промежуточный вывод</a></span></li></ul></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Модель-до-преобразования" data-toc-modified-id="Модель-до-преобразования-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Модель до преобразования</a></span></li><li><span><a href="#Модель-после-преобразования" data-toc-modified-id="Модель-после-преобразования-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Модель после преобразования</a></span></li><li><span><a href="#Промежуточный-вывод" data-toc-modified-id="Промежуточный-вывод-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Промежуточный вывод</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

# Описание данных

Признаки: пол, возраст и зарплата застрахованного, количество членов его семьи.

Целевой признак: количество страховых выплат клиенту за последние 5 лет.

## Загрузка данных

### Библиотеки

In [1]:
import pandas as pd
import numpy as np
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

### Подготовка данных

In [2]:
def analyze_dataframe(df):
    display("Первые строчки:", df.head())
    display("Последние строчки:", df.tail())
    display("Случайные строчки:", df.sample(5))
    display("Информация о каждом столбце:", df.describe())
    display("Общая информация:")
    df.info()
    display("Явные дубликаты:", df.duplicated().sum())
    display("Сортировка и количество пропущенных значений:", df.isnull().sum().sort_values())

In [3]:
data = pd.read_csv('/datasets/insurance.csv')
analyze_dataframe(data)

'Первые строчки:'

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

'Последние строчки:'

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
4995    0     28.0   35700.0            2                  0
4996    0     34.0   52400.0            1                  0
4997    0     20.0   33900.0            2                  0
4998    1     22.0   32700.0            3                  0
4999    1     28.0   40600.0            1                  0

'Случайные строчки:'

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
151     1     19.0   29300.0            0                  0
3866    1     32.0   23100.0            1                  0
281     1     39.0   48100.0            1                  0
2266    1     21.0   29500.0            0                  0
101     0     49.0   42500.0            1                  2

'Информация о каждом столбце:'

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

'Общая информация:'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


'Явные дубликаты:'

153

'Сортировка и количество пропущенных значений:'

Пол                  0
Возраст              0
Зарплата             0
Члены семьи          0
Страховые выплаты    0
dtype: int64

Удалим явные дубликаты 

In [4]:
data = data.drop_duplicates()

Изменим тип данных для столбцов Возраст и Зарплата

In [5]:
data['Возраст'] = data['Возраст'].astype(int)
data['Зарплата'] = data['Зарплата'].astype(int)

В итоге данные выглядт так

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4847 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   Пол                4847 non-null   int64
 1   Возраст            4847 non-null   int64
 2   Зарплата           4847 non-null   int64
 3   Члены семьи        4847 non-null   int64
 4   Страховые выплаты  4847 non-null   int64
dtypes: int64(5)
memory usage: 227.2 KB


### Промежуточный вывод

1. Последовательность столбцов логична.

2. В данных есть явные дубликаты - были удалены.

3. Изменён тип данных для столбц Возраст и Зарплата на целочисленный.

4. Пропуски в данных отсутствуют.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

### Задача

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Вопрос:** Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии? (Её можно обучить заново.)

### Решение

**Ответ:** Не изменится 

**Обоснование:** 

Нам нужно доказать, что 

$a=a'$

То есть доказать, что 

$w' = P^{-1}w$

Используя матричные свойства докажем

$ w' = ((XP)^T XP)^{-1} (XP)^Ty = (P^T X^T X P)^{-1} (P^T X^T) y = (P^T (X^T X) P)^{-1} (P^T X^T) y = P^{-1} (P^T (X^T X))^{-1} P^T X^T y = P^{-1} (X^T X)^{-1} (P^T)^{-1} P^T X^T y = //(P^T)^{-1} P^T = E// = P^{-1} (X^T X)^{-1} E X^Ty = P^{-1} (X^T X)^{-1} X^Ty = //(X^T X)^{-1} X^Ty = w // = P^{-1}w$

Что и требовалось доказать

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>



Успех 👍:




Самая частая ошибка - не учитывают что обратная матрица может быть только у квадратной.  Часто не соблюдают правила матричного преобразования, или меняют матрицы местами при перемножении как цифры.  


У тебя все ОК. 
    
    
    
<div class="alert alert-warning">



Совет 🤔:    
    
    
  


  
Есть ещё один способ доказательства   когда мы берём
    
    
$$
w = \arg\min_w MSE(Xw, y)
$$


(кстати можешь объяснить словами что это запись означает?)    
  
и     
    
$$
w_P = \arg\min_w MSE(XPw_p, y)
$$  
    
    
И делая  замену    w_p, видим что задача  
    
$$
\arg\min_w MSE(XPw_p, y)
$$
    
сводится к     
  
    
$$
\arg\min_w MSE(Xw, y)
$$    


Осталось только догадаться какую замену мы делаем.  Можешь попробовать решить и таким способом  

  
</div>
    

    




<div class="alert alert-info">
<h2> Комментарий студента: <a class="tocSkip"> </h2>

<b></b> Запись означает примерно следующее: w - оптимальный вес, который минимизирует среднеквадратичную ошибку между предсказанными и реальными значениями. 
    
Решение задачи выглядит как-то просто, сложно так-то сразу понять, что за замена тут
</div> 

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Успех 👍:



👍




<div class="alert alert-warning">


Совет 🤔:


По поводу 2 способа доказательства - оно очень короткое )    
    
    
    
    
Для исходных данных мы решаем это  
    
$$
w = \arg\min_w MSE(Xw, y)
$$
    
  
Для преобразованных данных решаем это  
    
$$
w_P = \arg\min_w MSE(XPw_p, y)
$$  
    
    
Давай сделаем замену   w_p на $P^{-1}w$ (Конечно мы это соотношение не берём из первого доказательства, иначе смысл во втором доказательстве просто нет. Тут типа математическая интуиция сработала, догадались как то о такой замене. Хотя догадаться можно, ведь видно что нам нужно избавиться от P)
    
 
И подставим:    
    
$$
\arg\min_w MSE(XPP^{-1}w, y)
$$
  
    
    
сократим и получим 
  
    
$$
\arg\min_w MSE(Xw, y)
$$    

    
    
    
    
Получается если заменить $w_p$ на $P^{-1}w$, то получаем ту же самую постановку задачи!  
     
    
</div>


### Промежуточный вывод

Теоретически доказано: перемножение признаков на обратимую матрицу даст тот же результат. Качество линейной регрессии не изменится.

## Алгоритм преобразования

**Алгоритм**

1. Создать случайную матрицу $P$ размерноcтью $(n,n)$ где n - количество признаков в датасете

2. Проверяем $P$ на обратимость:

    1)Если матрица обратима, идём дальше.

    2) Если матрица необратима, генерируем новые матрицы, до тех пор пока не получим обратимую матрицу или пока не достигнем лимита итераций.
    
2. Сохраняем матрицы $P$ и $P^{-1}$ для шифровки/дешифровки

3. Умножаем матрицу $X$ на $P$


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех 👍:


Принято

<div class="alert alert-warning">

Совет 🤔:

    
    

    
А если сгенерированная матрица будет необратимой? Тут можно расписать порядок действий на все случаи в том числе на этот. Кроме того можно написать как выбираем размер генерируемой матрицы. И тогда алгоритм будет полнее
    

<div class="alert alert-info">
<h2> Комментарий студента: <a class="tocSkip"> </h2>

<b></b> Согласен, подправил.
</div> 

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Успех 👍:



👍




<div class="alert alert-warning">




Совет 🤔:



Да, можно  генерировать матрицы до победного - это такой Инженерный подход.  А ещё можно взять единичную матрицу и применить  элементарный преобразования. Также можно использовать [матрицу по принципу диагонального доминирования](https://question-it.com/questions/13911024/generatsija-obratimyh-matrits-v-numpytensorflow) . Это уже  более математические способы 



**Обоснование**

$PP^{-1} = E$

$XPP^{-1} = X$

## Проверка алгоритма

### Модель до преобразования

Разделеним выборки на тестовую и тренировочную

In [7]:
features = data.drop(columns='Страховые выплаты', axis=1)
target = data['Страховые выплаты']
features_train, features_test, target_train, target_test = train_test_split(features,
                                                                            target,
                                                                            random_state=12345,shuffle=True,test_size=0.25)

Опишем Линейную регрессию

In [8]:
class LinearRegression:
    def fit(self, train_features, train_target):
        X = np.concatenate((np.ones((train_features.shape[0], 1)), train_features), axis=1)
        y = train_target
        X_transpose = X.T
        XTX_inv = np.linalg.inv(X_transpose.dot(X))
        w = XTX_inv.dot(X_transpose).dot(y)
        self.w = w[1:]
        self.w0 = w[0]
        
    def predict(self, test_features):
        return test_features.dot(self.w) + self.w0
    

In [9]:
model = LinearRegression()
model.fit(features_train, target_train)
predictions = model.predict(features_test)
print('R2:',r2_score(target_test, predictions))

R2: 0.423077276158367


### Модель после преобразования

Создадим матрицу $P$

In [10]:
def get_rand_matrix():
    det = 0
    while det == 0:
        matrix = np.random.randint(12345, size=(features.shape[1], features.shape[1]))
        det = np.linalg.det(matrix)
    return matrix

In [12]:
p_matrix = get_rand_matrix()
p_matrix

array([[ 1725,  7625,  7025,   700],
       [ 6047,  7703,  5901,  2182],
       [11966,  8972,  8908,  5347],
       [ 5175,  6221,  2856,  5429]])

Проверим матрицу на обратимость, создадим матрицу $P^{-1}$




<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>

Совет 🤔:



- ты проверяешь    матрицу на обратимость - отлично, но если она необратимая код до конца не исполнится
 


   
Можно здесь было сделать похитрей - генерируем матрицу пока не получим обратимую и наш код  гарантированно полностью отработает. Добиться этого можно например так:


    
    def get_rand_matrix():
        det = 0
        while det == 0:
            matrix = np.random.normal(size=(data.shape[1], data.shape[1]))
            det = np.linalg.det(matrix)
        return matrix

C помощью рекурсии

    def func(dim):
        transform_matrix = np.random.rand(dim, dim)
        try: 
            inv = np.linalg.inv(transform_matrix)
            return transform_matrix, inv
        except:
            func(dim)  
    

- что насчет универсальности кода? Сейчас ты цифрой определяешь размерность случайной матрички, почему бы не написать код, который отработает для любой размерности матрицы обьект-признак?      
    



<div class="alert alert-info">
<h2> Комментарий студента: <a class="tocSkip"> </h2>

<b></b> Согласен, подправил. На счёт универсального кода, а твой код:
    
    
    
    def get_rand_matrix():
        det = 0
        while det == 0:
            matrix = np.random.normal(size=(data.shape[1], data.shape[1]))
            det = np.linalg.det(matrix)
        return matrix

Не будет универсальным? Только нужно подправить data изменить на features. А .normal на .randit     
</div> 

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Совет 🤔:



    Не будет универсальным? 

    
Будет, я ведь не цифрой определяю размер матрицы а вытаскиваю его из датасета    

Посмотрим на данные после переменожения

In [14]:
multiple_feature = features @ p_matrix
multiple_feature.columns = features.columns
multiple_feature.head()

Пол    Возраст   Зарплата  Члены семьи
0  593768427  445340869  442088622    265306791
1  454991337  341296559  338778302    203291801
2  251461363  188635387  187239129    112350278
3  499119537  374306605  371593233    223026580
4  312483641  234392509  232671053    139618496

Теперь вместо данных о клиенте - непонятные числа

Перемножим выборки с матрицей $P$ и обучим модель Линейной регрессии

In [15]:
features_train = features_train.dot(p_matrix)
features_test = features_test.dot(p_matrix)

model = LinearRegression()
model.fit(features_train, target_train)
predictions = model.predict(features_test)
print('R2:',r2_score(target_test, predictions))

R2: 0.4230772217865213


### Промежуточный вывод

1. Точность от применения преобразования не изменилась, качество модели до преобразования R2: 0.423077276158367 качество модели после преобразования R2: 0.42307712437097766.



<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех 👍:
    
Все сошлось. Отдельный плюс за: 
    
    


- использование train_test_split - все правильно, помним о стандартах ML при моделировании



- использование class LinearRegression, а не готовую библиотечку, где все скрыто, а так все видно и наглядно что полезно при обучении




## Вывод

В ходе проекта мы защитили данные клиентов. 

Проверили теоретически и практически, что качество линейной регресии не изменилось от защиты.

В ходе работы мы выполнили следующие шаги:

1. Загрузка данных:

    1. Последовательность столбцов логична.

    2. В данных есть явные дубликаты - были удалены.

    3. Изменён тип данных для столбц Возраст и Зарплата на целочисленный.

    4. Пропуски в данных отсутствуют.
    
2. Умножение матриц
    
    Теоретически доказано: перемножение признаков на обратимую матрицу даст тот же результат. Качество линейной регрессии не изменится.
    
    
3. Алгоритм преобразования

    Описали и обасновали алгоритм преобразования
    
    
4. Проверка алгоритма

    Практически доказано: Точность от применения преобразования не изменилась, качество модели до преобразования R2: 0.423077276158367 качество модели после преобразования R2: 0.42307712437097766.


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех 👍:


Илья, здорово что в конце расписана все этапы работы. Это важно потому что когда проект захочет посмотреть будущий работодатель (или начальник), у него может не быть времени на подробный разбор кода. Вероятнее всего он бегло просмотрит код, а из общего вывода захочет получить представление о всей работе.



<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюера</b></font>


Илья, у тебя старательно выполненная работа, все четко, осмысленно. Отмечаю уровень твоей программистской подготовки. 

    
    

Я оставил небольшие советы и вопросики (если есть время и желание можешь воспользоваться/ответить)
    
    
    
    

Посылаю на вторую итерацию, чтобы ответить на вопросы если они появятся. На связи
    









<div class="alert alert-warning">
<b>Комментарий ревьюера</b>



На мой взгляд проект очень необычный, выделяющийся из остальных проектов потому что касается раздела математики лежащей в основе машинного обучения  (помимо производных, матанализа и теории вероятности с матстатистикой). Если в дальнейшем ты углубишься в изучении нейростетей, то понимание линейной алгебры будет очень полезным, потому что, по сути, все нейросети - это перемножение матриц и Линейная алгебра даёт очень компактный способ записи того что происходит в нейросетях. 

Кроме того языком линейной алгебра можно объяснить тематическое моделирование и pca (через сингулярное разложение)

Мне не очень нравится эти ссылки, наверняка можно найти получше, но там вполне наглядно показывают связь между линейной алгебры и многослойным перцептроном
    
https://www.youtube.com/watch?v=kmdqBlls_HQ    
https://www.youtube.com/watch?v=AOZZeTQl4Xc


Это даже важно как "понятийный язык", вот здесь о модели word2vec (это NLP) рассказываться "языком линейной алгебры" с [22 минуты](https://www.youtube.com/watch?v=MBQdMQUZMQM&t=3340s) 
    

Возможно будет полезно:
    
    

    
- Если хочешь посмотреть вообще про линейную алгебру и матрицы, то мне нравится  вот это [видео](https://www.youtube.com/watch?v=RNTRYicPvWQ&list=PLVjLpKXnAGLXPaS7FRBjd5yZeXwJxZil2), рассказывается по существу и увлекательно.  


- по применению криптографии в машинном обучении на английском: [тут](https://www.cs.cmu.edu/~rjhall/JOS_revised_May_31a.pdf)




<div class="alert alert-info">
<h2> Комментарий студента: <a class="tocSkip"> </h2>

<b></b> Спасибо за ревью! Спасибо за материал!
</div> 


<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюераV2</b></font>

Спасибо за работу!    

    

Красного нет, вопросов нет (на те что были ответил), значит все, пора принимать) Надеюсь мои советы и вопросики были полезны и в копилочку знаний упало что то новое, а проект стал лучше, и симпатичней.

  
Отличная работа  Илья. Желаю успехов в дальнейшей учебе!


## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования